lib/sshutils/sftp/sftp.go

```diff
@@ -138,6 +143,87 @@ func CreateDownloadConfig(src, dst string, opts Options) (*Config, error) {
 	return c, nil
 }
 
+// HTTPTransferRequest describes file transfer request over HTTP.
+type HTTPTransferRequest struct {
+	// Dst is the source file name
+	Src string
+	// Dst is the destination file name
+	Dst string
+	// HTTPRequest is where the source file will be read from for
+	// file upload transfers
+	HTTPRequest *http.Request
+	// HTTPResponse is where the destination file will be written to for
+	// file download transfers
+	HTTPResponse http.ResponseWriter
+}
+
+// CreateHTTPUploadConfig returns a Config ready to upload a file over
+// HTTP.
+func CreateHTTPUploadConfig(req HTTPTransferRequest) (*Config, error) {
+	if req.Src == "" {
+		return nil, trace.BadParameter("source path is empty")
+	}
+	if req.Dst == "" {
+		return nil, trace.BadParameter("destination path is empty")
+	}
+	if req.HTTPRequest == nil {
+		return nil, trace.BadParameter("HTTP request is empty")
+	}
+
+	contentLength := req.HTTPRequest.Header.Get("Content-Length")
+	fileSize, err := strconv.ParseInt(contentLength, 10, 0)
+	if err != nil {
+		return nil, trace.Errorf("failed to parse Content-Length header: %w", err)
+	}
+
+	c := &Config{
+		srcPaths: []string{req.Src},
+		dstPath:  req.Dst,
+		srcFS: &httpFS{
+			reader:   req.HTTPRequest.Body,
+			fileName: req.Src,
+			fileSize: fileSize,
+		},
+		dstFS: &localFS{
+			partOfHTTPTransfer: true,
+		},
+		isHttpTransfer: true,
+	}
+	c.setDefaults()
+
+	return c, nil
+}
+
+// CreateHTTPDownloadConfig returns a Config ready to download a file
+// over HTTP.
+func CreateHTTPDownloadConfig(req HTTPTransferRequest) (*Config, error) {
+	if req.Src == "" {
```

Comment:
This code is duplicated above. How about adding a checkandsetdefaults?

Reply:
Good idea, done